File: macos/Sources/TypedTextBuffer.swift
import Foundation

// Bounded keystroke-only buffer, including standard Korean two-set composition.
final class TypedTextBuffer {
    private var committed = ""
    private var jamo: [Character] = []
    private static let leads = Array("ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ")
    private static let vowels = Array("ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ")
    private static let tails = Array(" ㄱㄲㄳㄴㄵㄶㄷㄹㄺㄻㄼㄽㄾㄿㅀㅁㅂㅄㅅㅆㅇㅈㅊㅋㅌㅍㅎ")
    private static let keys = Array("rRseEfaqQtTdwWczxvgkoiOjpuPhynbml")
    private static let letters = Array("ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎㅏㅐㅑㅒㅓㅔㅕㅖㅗㅛㅜㅠㅡㅣ")
    var text: String { String((committed + Self.compose(jamo)).suffix(240)) }
    func clear() { committed = ""; jamo.removeAll(keepingCapacity: true) }
    func commit() {
        committed = String((committed + Self.compose(jamo)).suffix(240))
        jamo.removeAll(keepingCapacity: true)
    }
    func append(_ value: String, korean: Bool) {
        for character in value {
            let spelling = String(character)
            let key = "REQTWOP".contains(character) ? character : (spelling.lowercased().first ?? character)
            if korean, let index = Self.keys.firstIndex(of: key) {
                jamo.append(Self.letters[index])
                if jamo.count > 1024 { jamo.removeFirst(256) }
            } else {
                commit()
                committed = String((committed + spelling).suffix(240))
            }
        }
    }
    func backspace() {
        if !jamo.isEmpty { jamo.removeLast() }
        else if !committed.isEmpty { committed.removeLast() }
    }
    private static func compoundVowel(_ a: Int, _ b: Int) -> Int? {
        switch (a, b) {
        case (8, 0): return 9
        case (8, 1): return 10
        case (8, 20): return 11
        case (13, 4): return 14
        case (13, 5): return 15
        case (13, 20): return 16
        case (18, 20): return 19
        default: return nil
        }
    }
    private static func compoundTail(_ a: Int, _ b: Int) -> Int? {
        if a == 1 && b == 19 { return 3 }
        if a == 4 && b == 22 { return 5 }
        if a == 4 && b == 27 { return 6 }
        if a == 8, let index = [1, 16, 17, 19, 25, 26, 27].firstIndex(of: b) { return 9 + index }
        return a == 17 && b == 19 ? 18 : nil
    }
    private static func splitTail(_ value: Int) -> (Int, Character) {
        switch value {
        case 3: return (1, "ㅅ")
        case 5: return (4, "ㅈ")
        case 6: return (4, "ㅎ")
        case 9...15: return (8, Array("ㄱㅁㅂㅅㅌㅍㅎ")[value - 9])
        case 18: return (17, "ㅅ")
        default: return (0, tails[value])
        }
    }
    private static func compose(_ source: [Character]) -> String {
        var output = "", lead = -1, vowel = -1, tail = 0
        func flush(_ l: Int, _ v: Int, _ t: Int) {
            if l >= 0 && v >= 0 { output.unicodeScalars.append(UnicodeScalar(0xAC00 + (l * 21 + v) * 28 + t)!) }
            else if l >= 0 { output.append(leads[l]) }
            else if v >= 0 { output.append(vowels[v]) }
        }
        for character in source {
            if let nextVowel = vowels.firstIndex(of: character) {
                if vowel < 0 { vowel = nextVowel }
                else if tail > 0 {
                    let (first, second) = splitTail(tail)
                    flush(lead, vowel, first)
                    lead = leads.firstIndex(of: second) ?? -1; vowel = nextVowel; tail = 0
                } else if let combined = compoundVowel(vowel, nextVowel) { vowel = combined }
                else { flush(lead, vowel, tail); lead = -1; vowel = nextVowel; tail = 0 }
            } else if let nextLead = leads.firstIndex(of: character) {
                let nextTail = tails.firstIndex(of: character) ?? -1
                if lead >= 0 && vowel >= 0 && nextTail > 0 {
                    if tail == 0 { tail = nextTail }
                    else if let combined = compoundTail(tail, nextTail) { tail = combined }
                    else { flush(lead, vowel, tail); lead = nextLead; vowel = -1; tail = 0 }
                } else { flush(lead, vowel, tail); lead = nextLead; vowel = -1; tail = 0 }
            }
        }
        flush(lead, vowel, tail)
        return output
    }
}

File: src/TypedTextBuffer.cs
using System;
using System.Text;

namespace EmotionCat
{
    // Only characters supplied by keyboard events enter this bounded RAM buffer.
    // A two-set Hangul run is recomposed so backspace and final-consonant splits
    // work without asking another application to expose an editable text field.
    internal sealed class TypedTextBuffer
    {
        readonly StringBuilder committed = new StringBuilder();
        readonly StringBuilder jamo = new StringBuilder();
        const string Leads = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ";
        const string Vowels = "ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ";
        const string Tails = " ㄱㄲㄳㄴㄵㄶㄷㄹㄺㄻㄼㄽㄾㄿㅀㅁㅂㅄㅅㅆㅇㅈㅊㅋㅌㅍㅎ";
        const string Keys = "rRseEfaqQtTdwWczxvgkoiOjpuPhynbml";
        const string Letters = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎㅏㅐㅑㅒㅓㅔㅕㅖㅗㅛㅜㅠㅡㅣ";

        public string Text { get { return Tail(committed.ToString() + Compose(jamo.ToString()), 240); } }
        public void Clear() { committed.Clear(); jamo.Clear(); }
        public void Commit() { committed.Append(Compose(jamo.ToString())); jamo.Clear(); Trim(); }
        public void Append(string text, bool korean)
        {
            foreach (char c in text)
            {
                char key = c;
                if (Char.IsUpper(key) && "REQTWOP".IndexOf(key) < 0) key = Char.ToLowerInvariant(key);
                int index = Keys.IndexOf(key);
                if (korean && index >= 0)
                {
                    jamo.Append(Letters[index]);
                    // Keep at most 1024 uncommitted physical jamo; output is 240 chars.
                    if (jamo.Length > 1024) jamo.Remove(0, 256);
                }
                else if (!Char.IsControl(c)) { Commit(); committed.Append(c); Trim(); }
            }
        }
        public void Backspace()
        {
            if (jamo.Length > 0) jamo.Length--;
            else if (committed.Length > 0)
            {
                int count = committed.Length > 1 && Char.IsLowSurrogate(committed[committed.Length - 1]) && Char.IsHighSurrogate(committed[committed.Length - 2]) ? 2 : 1;
                committed.Length -= count;
            }
        }
        void Trim() { if (committed.Length > 240) { string text = Tail(committed.ToString(), 240); committed.Clear(); committed.Append(text); } }
        static string Tail(string s, int n) { int start = Math.Max(0, s.Length - n); if (start > 0 && Char.IsLowSurrogate(s[start])) start++; return s.Substring(start); }
        static int CompoundVowel(int a, int b)
        {
            if (a == 8) { if (b == 0) return 9; if (b == 1) return 10; if (b == 20) return 11; }
            if (a == 13) { if (b == 4) return 14; if (b == 5) return 15; if (b == 20) return 16; }
            return a == 18 && b == 20 ? 19 : -1;
        }
        static int CompoundTail(int a, int b)
        {
            if (a == 1 && b == 19) return 3;
            if (a == 4) { if (b == 22) return 5; if (b == 27) return 6; }
            if (a == 8) { int i = Array.IndexOf(new[] { 1, 16, 17, 19, 25, 26, 27 }, b); if (i >= 0) return 9 + i; }
            return a == 17 && b == 19 ? 18 : -1;
        }
        static void SplitTail(int t, out int first, out char second)
        {
            first = 0; second = Tails[t];
            if (t == 3) { first = 1; second = 'ㅅ'; }
            else if (t == 5 || t == 6) { first = 4; second = t == 5 ? 'ㅈ' : 'ㅎ'; }
            else if (t >= 9 && t <= 15) { first = 8; second = "ㄱㅁㅂㅅㅌㅍㅎ"[t - 9]; }
            else if (t == 18) { first = 17; second = 'ㅅ'; }
        }
        static void Flush(StringBuilder output, int l, int v, int t)
        {
            if (l >= 0 && v >= 0) output.Append((char)(0xAC00 + (l * 21 + v) * 28 + t));
            else if (l >= 0) output.Append(Leads[l]);
            else if (v >= 0) output.Append(Vowels[v]);
        }
        internal static string Compose(string source)
        {
            var output = new StringBuilder(); int l = -1, v = -1, t = 0;
            foreach (char c in source)
            {
                int nextVowel = Vowels.IndexOf(c), nextLead = Leads.IndexOf(c), nextTail = Tails.IndexOf(c);
                if (nextVowel >= 0)
                {
                    if (v < 0) v = nextVowel;
                    else if (t > 0)
                    {
                        int first; char second; SplitTail(t, out first, out second);
                        Flush(output, l, v, first); l = Leads.IndexOf(second); v = nextVowel; t = 0;
                    }
                    else
                    {
                        int compound = CompoundVowel(v, nextVowel);
                        if (compound >= 0) v = compound;
                        else { Flush(output, l, v, t); l = -1; v = nextVowel; t = 0; }
                    }
                }
                else if (nextLead >= 0)
                {
                    if (l >= 0 && v >= 0 && nextTail > 0)
                    {
                        if (t == 0) t = nextTail;
                        else
                        {
                            int compound = CompoundTail(t, nextTail);
                            if (compound >= 0) t = compound;
                            else { Flush(output, l, v, t); l = nextLead; v = -1; t = 0; }
                        }
                    }
                    else { Flush(output, l, v, t); l = nextLead; v = -1; t = 0; }
                }
            }
            Flush(output, l, v, t); return output.ToString();
        }
    }
}
